metsim/metsim.py

```diff
@@ -263,34 +263,41 @@ def launch(self):
         time_dim, groups = self._get_time_dim_and_time_dim()
 
         for label, times in groups.items():
-            self.pool = Pool(processes=nprocs)
-            self.site_generator = itertools.product(*iter_list)
+            self.site_generator = list(itertools.product(*iter_list))
             logger.info("Beginning {}".format(label))
-            status = []
 
             # Add in some end point data for continuity in chunking
             times_ext = times.union([times[0] - pd.Timedelta("1 days"),
                                      times[-1] + pd.Timedelta("1 days")]
                                     ).intersection(time_dim)
             data = self.met_data.sel(time=times_ext)
             self.setup_output(self.met_data.sel(time=times))
-            for site in self.site_generator:
+
+            run_func = dask.delayed(wrap_run, name='metsim_run', pure=True,
+                                    traverse=False, nout=3)
+            delayed_objs = []
+
+            for site in progress(self.site_generator):
                 locs = {k: v for k, v in zip(self.params['iter_dims'], site)}
                 # Don't run masked cells
                 if not self.domain['mask'].sel(**locs).values > 0:
                     continue
-                stat = self.pool.apply_async(
-                    wrap_run,
-                    args=(self.method.run, locs, self.params,
-                          data.sel(**locs),
-                          self.state.sel(**locs),
-                          self.disagg, times, label),
-                    callback=self._unpack_results)
-                status.append(stat)
-            self.pool.close()
-            # Check that everything worked
-            [stat.get() for stat in status]
-            self.pool.join()
+                delayed_objs.append(run_func(self.method.run, locs,
+                                             self.params,
+                                             data.sel(**locs),
+                                             self.state.sel(**locs),
+                                             self.disagg,
+                                             times, label).persist())
```

Comment:
I'm considering unpacking the inputs to this function from their xarray data structures into something simpler (dict of numpy arrays). I'm hypothesizing that this will make it easier to serialize the data when using multiprocessing. @arbennett, thoughts?

Reply:
My simple tests indicate that this may yield about a 2x speedup.

Reply:
This seems like a good idea to me.